### PR TITLE
Prototype for standard basis for ideals and modules in Oscar

### PR DESCRIPTION
**Pull request by HechtiDerLachs**
Whether or not a proper new type for standard basis of ideals and modules should be introduced in Oscar is an ongoing debate. With this PR we would like to contribute to that discussion by exposing our specific demands and suggestions for solutions.

Let me first explain, why @afkafkafk13 and I are interested in a standard-basis functionality. This July, we would like to give a course on computational aspects of singularity theory at the [Sao Carlos workshop](http://www.worksing.icmc.usp.br/main_site/2022/) which is one of the major international conferences in singularity theory. Two years ago, we would have given the course based on Singular, but this year, we would very much like to use and promote Oscar; especially for the upcoming generation of singularity theorists. 

### An example problem we would like to address
Now we are facing the following challenges. The compuational aspects of singularity theory rely heavily on the theory of standard basis and monomial orderings. As a central example, let us consider the Kodaira spencer map `K : T_0 B -> T^1_{X,0}`. For a given singularity `(X, 0)` and a deformation of it `(Z, 0) -> (B,0)` over some base germ `(B,0)` it takes a tangent vector `v` in the tangent space at `0` of the base `B` to its corresponding infinitesimal deformation of the singularity. The space `N := T^1_{X,0}` will be probably be a `SubQuo`, but certainly a finitely presented module over a polynomial ring `R`, which has finite dimension over the field `k` of coefficients of `R`. For simplicity, let us assume for the moment that `N = F/M` for some free module `F = R^n` and a submodule `M` of `F`.

In order to write down `K`, we need a so-called `k`-base of `N` with respect to a local ordering. This is a finite list of representatives `v_1,...,v_r` in `F` of elements of `N` which reduce to a basis for `N` as a vector space over `k`. On the mathematical side, this is achieved by 

 1) Choosing a (local) monomial ordering `<` for `F`;
 2) computing a standard basis `G` for `M` w.r.t. `<` and
 3) collecting the monomials (in the module sense) which are not contained in the leading module of `G`.

 This is one of the things we would like to explain to the students and accompany this with a presentation of the corresponding commands and computations in Oscar. 

### Why not use Singular?
The methods for all the above exist since many years in Singular and they are available through the calls to the libraries via `Singular.jl`. But when teaching a course, there is no point in using Oscar, only to tell the students that, essentially, we will be working with Singular, including the use of the singular syntax etc. What we would like to have, are the Oscar versions of commands like `k`-base, etc., meaning we use Oscar rings and modules and the return value of `k_base` is not only a list of elements, but an honest `Oscar` vector space `V` together with a `Hecke.Map` taking a basis of `V` to the corresponding representatives of the quotient `N = F/M`. 

### What could standard bases look like in Oscar?
Our point of view is the following: Groebner- and standard basis are not only concrete objects in `Singular`, but honest mathematical objects with a mathematical theory of its own right. We think, this should be reflected in Oscar by introducing the corresponding abstract types and interfaces; a prototype of this can be found in this PR in `experimental/standard_basis.jl`. 

In particular, a standard basis `G` is more than just a list of elements; for instance, the ordering with respect to which a basis had been computed needs to be associated to `G`. Furthermore, having an abstract and concrete types for standard bases allows us to write methods like e.g. `normal_form` which take an instance  `<:AbsStandardBasis` as input rather than simply a `Vector{T}` where `{T<:MPolyElem}`. Just to compare: For factorizations of polynomials, we have a type of its own for the return value of `factor(f)`, namely `AbstractAlgebra.Fac`. So why not have a type for standard basis?

### User facing vs. programmer facing functionality, code organization
Introducing a type for standard bases should also help us to improve the organization of our code.  In the end, we want to be able to not only use Singular as a backend for standard basis computations, but also any other potential and upcoming Groebner-basis engines. Having an abstract interface allows us to embed and organize the concrete implementations of standard bases for different backends. 

Keep in mind that Groebner- and standard bases should not be user-facing for the average user, in the end. But we believe that they should be 'programmer facing' in the sense that there should be a second layer below the surface, to which a programmer can refer and this would be the layer, we would like to expose to our students in the course. At this point, this second layer is concretely given by objects like `BiPolyArray` or `ModuleGens`, accompanied with their methods `singular_assure` and `oscar_assure`. But this is more a technical and practical solution to allow for things like ideal membership tests to be carried out for Oscar rings. It is **not** the solution, we would like to present to students in a course! In a course, we introduce the mathematical theory of testing ideal membership via standard basis methods and then show how to do that in Oscar. If showing the latter means introducing the students to `BiPolyArrays` and `singular_assure`, something is fundamentally wrong with the course! 

### Conclusion
For the preparation of the course, @afkafkafk13 and I need to have a reliable decision on the design of these matters within Oscar and, I'm afraid, we need that rather soon. 

Should we reach the conclusion that Oscar should feature standard basis as mathematical objects of their own right, then we will be happy to contribute to their development.

Should we decide that we stick with the practical solution via `BiPolyArrays`, then it seems that Oscar is not a suitable software for this course. But that would entail that we miss the opportunity to promote it at this conference.

Please, share your thoughts, comments, suggestions, complaints, ideas and objections.

CC: @wdecker , @ederc , @jankoboehm , @AlexD97 , @fingolfin

### CODE REVIEW

**Comment by thofma** [comment]
Hm things like
```
### return the leading module of G
function leading_module(G::AbsModuleStdBasis)
  # can be implemented generically
end

### return the dimension of the support of the module
function dim(G::AbsModuleStdBasis)
end

function is_finite_dimensional(G::AbsModuleStdBasis) 
end

function as_vector_space(G::AbsModuleStdBasis)
```
are kind of the reasons why this would be bad. Interpreting bases as modules is exactly what we don't want to do or encourage.

**Comment by afkafkafk13** [comment]
For dimension and is_finite_dimensional I see your point. These notions are independent of the chosen ordering and hence a bad example.

But leading_module for obtaining the module generated by the leading terms is a genuine functionality of its own associating a closely related module to a set of generators (making the first step towards combinatorial approaches within the monomial diagram), but not to the module generated by them -- as the leading module depends on the ordering. It is this step from the ideal/module to combinatorial data (sufficient for many purposes) which is widely used in singularity theory. Obviously, this should not be exposed to the novice or occasional user, but it should be there for more sophisticated use in home-made programs.

Also for the map relating the (a priori abstractly defined) k-vector space to the 0-dimensional quotient of a free module by a submodule is depending heavily on the chosen ordering as this determines the choice of representatives of the images of the basis  -- and very often users want some explicit choice of representatives (be it that they know which choice or be it that they accept any sensible automated choice). -- The same holds for graded slices in graded modules.

So yes, we should go through our list again and see, which data is independent of the chosen ordering and use the ideal and module instead of the standard basis in these instances, but there are still important instances left, where we need to work with a standard basis w.r.t. a fixed ordering. 

Yet again, working with representatives and explicit choices of representatives is natural to singularity theory and we should have a readable way to support this without saying singular.___  or BiPolyArray anywhere on the designated level for the sophisticated user/occasional programmer. The latter commands are one level deeper in the code and should only be used by people who know what they are doing and what they want to force to be done on the OSCAR side or the Singular side.

**Comment by fingolfin** [comment]
Is the only concern with `BiPolyArray` its name?! Otherwise, isn't it made exactly for these kind of things, i.e. a list of Oscar polynomial generators with an (optional) underlying Singular Gröbner basis?

**Comment by HechtiDerLachs** [comment]
I agree that some of these points are valid. Also, I've just had a discussion with @thofma and became aware of the following fact: 

A purely mathematical Groebner basis `G` of type, say, `MathGB` does not have any further functionality than 

 1) `ordering(G)`
 2) `elements(G)`
 3) `ambient_free_module(G)`

But what we need for a functioning workhorse is more: A set of original generators, a transition matrix for these generators to the Groebner basis, and, last but not least, something like a 'Singular side' in the cache where copies of the Oscar polynomials are cached. In particular, if the latter didn't exist, then any use of `G` in `normal_form` would produce the overhead of conversion again. 

Therefore, purely mathematical Groebner bases can not be functional for use in Oscar. This contradicts my hope that introducing a Groebner basis interface could be a 'dual use' case where on the one hand we have something to present to the student, but on the other hand have an interface for computational backends and communicating with groebner engines. 

The conclusion seems to emerge from this is, that we indeed arrive at something that is more like an enhanced version of `BiPolyArray` rather than a mathematical Groebner basis. 

Whether or not introducing an abstract interface for such types makes sense is then another debate. But then @afkafkafk13 and I need to reconsider: The literature that we would like to base our course on is using the terminology of standard bases, etc. Is it possible or even reasonable to illustrate these concepts in Oscar?

**Comment by thofma** [comment]
(Just a comment regarding your original aim with the workshop. Something that I did in the past was to provide some additional things that were used during the workshop as follows. At the beginning everybody does `]add github.com/HechtiDerLachts/FancyNameWorkshop` and then `using FancyNameWorkshop`. This way one can add some helper functions to enhance Oscar or whatever. For example introducing a mathematical GB to mirror the text book.)

File: experimental/LibSingWrapper.jl [hunk 1]
```diff
@@ -0,0 +1,244 @@
...
+  WI = W(I)
+  # compute a standard basis w.r.t a local ordering 
+  stdI = groebner_basis(WI)
+
+  # carry out the computations on the Singular side
+  Singular.dimension(singular_gens(stdI)) == 0 || error("singularity is not isolated")
+  milnor_number = Singular.vdim(singular_gens(stdI)) 
+  # make the conversion to the Oscar side 
+  milnor_kbase = [to_oscar_side(stdI, g) for g in gens(Singular.kbase(singular_gens(stdI)))]
+  return (stdI, milnor_number, milnor_kbase)
```
Comment (fieker) [inline experimental/LibSingWrapper.jl:55]: In julia as opposed to Singular, you'd in general return three values not a triple. The use is the same, but notionally here and else where you have 3 return values, not a single triple. You'd want to use them independently I assume...

File: experimental/LibSingWrapper.jl [hunk 2]
```diff
@@ -0,0 +1,244 @@
...
+end
+
+### The global Milnor number of a hypersurface with at most isolated singularities
+@Markdown.doc """
+    global_milnor_number(f::MPolyElem)
+
+Computes the global Milnor number of the affine hypersurface 
+defined by `f` with at most isolated singularities.
+"""
+function global_milnor_number(f::MPolyElem{T}) where {T<:FieldElem}
```
Comment (fieker) [inline experimental/LibSingWrapper.jl:110]: Your choice, but looking at the names you should have
 - tjurina -> the number + data
 - tjurina_number
 - tjurina_number_global (then tab-expansion will find it)

File: experimental/singular_standard_basis.jl [hunk 3]
```diff
@@ -0,0 +1,73 @@
+mutable struct SingularStandardBasis{
```
Comment (fieker) [inline experimental/singular_standard_basis.jl:1]: I actually don't think you gain anything from having all that type-parameters - but you loose readability. (Maybe @thofma  has a different opinion?) The elem types might have a performance benefit, but unless the others are used for dispatch, why?

File: experimental/singular_standard_basis.jl [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+    end
+    R = base_ring(F)
+    return new{elem_type(R), typeof(F), typeof(generated_submodule), ModElemType, typeof(o)}(F, generated_submodule, o, v)
+  end
+end
+
+ordering(G::SingularStandardBasis) = G.o
+ambient_free_module(G::SingularStandardBasis) = G.F
+
+function sub(G::SingularStandardBasis; cached::Bool=true)
```
Comment (fieker) [inline experimental/singular_standard_basis.jl:45]: again, mixing module and basis. a sub of a basis, to me is a subset of the basis

File: experimental/singular_standard_basis.jl [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+
+ordering(G::SingularStandardBasis) = G.o
+ambient_free_module(G::SingularStandardBasis) = G.F
+
+function sub(G::SingularStandardBasis; cached::Bool=true) 
+  cached && return G.M, inclusion_map(M)
+  return sub(G.M, G.F)
+end
+    
+function sub(G::SingularStandardBasis; cached::Bool=true)
```
Comment (fieker) [inline experimental/singular_standard_basis.jl:50]: I am blind, but that is identical to line 45?

File: experimental/singular_standard_basis.jl [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+    end
+    N, p = quo(ambient_free_module(G), G.M)
+    G.quotient_module = N
+    G.quotient_projection = p
+    return N, p
+  end
+  return sub(G.M, G.F)
+end
+
+function elements(G::SingularStandardBasis)
```
Comment (fieker) [inline experimental/singular_standard_basis.jl:63]: is this, what elsewhere would be collect? Ie the contents as a vector/ iterator?

File: experimental/standard_basis.jl [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+function ordering(G::AbsIdealStdBasis)
+end
+  
+### return the ambient free module F which contains the elements of G
+function ambient_free_module(G::AbsModuleStdBasis)
+end
+
+### return the submodule M ⊂ F ≅ Rⁿ of which G is a standard basis, 
+# together with its inclusion M ↪ F.
+function sub(G::AbsModuleStdBasis; cached::Bool=true)
```
Comment (fieker) [inline experimental/standard_basis.jl:24]: mixing module and basis

File: experimental/standard_basis.jl [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+end
+
+### return the quotient F/M for the submodule M ⊂ F ≅ Rⁿ of which G 
+#is a standard basis, together with its projection F →  F/M.
+function quo(G::AbsModuleStdBasis; cached::Bool=true)
+end
+
+### return a vector of the module elements in the standard basis
+function elements(G::AbsModuleStdBasis)
```
Comment (fieker) [inline experimental/standard_basis.jl:33]: collect

File: experimental/standard_basis.jl [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+end
+
+### return the submodule M ⊂ F ≅ Rⁿ of which G is a standard basis, 
+# together with its inclusion M ↪ F.
+function sub(G::AbsModuleStdBasis; cached::Bool=true)
+end
+
+### return the quotient F/M for the submodule M ⊂ F ≅ Rⁿ of which G 
+#is a standard basis, together with its projection F →  F/M.
+function quo(G::AbsModuleStdBasis; cached::Bool=true)
```
Comment (fieker) [inline experimental/standard_basis.jl:29]: mixing

File: experimental/standard_basis.jl [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+### Essential functionality ###########################################
+
+function normal_form(v::FreeModElem, G::AbsModuleStdBasis)
+end
+
+### In the case of a finite dimensional quotient F/G with F ≅ Rⁿ 
+# a free module over a polynomial ring R = 𝕜[x₁,…,xₙ] with an ordering <, 
+# and 𝕜 a field, return a list of elements v₁,…,vᵣ ∈ F which reduce 
+# to a 𝕜-basis of F/G.
+function k_base(G::AbsModuleStdBasis)
```
Comment (fieker) [inline experimental/standard_basis.jl:50]: mixing

File: experimental/standard_basis.jl [hunk 11]
```diff
@@ -0,0 +1,108 @@
...
+function graded_slice(G::AbsModuleStdBasis, d::Int; w::Vector{T}=Vector{Int}()) where {T<:Integer}
+end
+
+
+### Generic functions #################################################
+
+base_ring(G::AbsModuleStdBasis) = base_ring(ambient_free_module(G))
+
+### return the leading module of G
+function leading_module(G::AbsModuleStdBasis)
```
Comment (fieker) [inline experimental/standard_basis.jl:69]: mixing

**Comment by HechtiDerLachs** [comment]
@fieker : Thanks for your comments. 

I understand that @afkafkafk13 and me were still pretty Singular-biased when we tried to formulate our framework. Indeed, care must be taken in order to not mix up the functionality for ideals and modules in Oscar with functionality for standard bases or even lists and vectors (as were used in singular). 

Probably, the hybrid object that we are looking for is not a mathematical standard basis, in the end, but more an Oscar object that we can use for 
 1) communicating the starndard-basis related commands 
 2) caching the intermediate results on the side of the groebner engine (be it singular or anything else)
 3) harvesting the outcoming pure Oscar-objects.

After all, that sounds more like an enhanced version of `BiPolyArray` and `ModuleGens`. And I was informed that something of that sort is currently in the making. 

Hence, I would suggest to postpone further discussion until we can have a look at such prototypes and how commands related to standard bases and orderings can be bound to that. Thanks again to everyone for participating up to this point.

**Pull request closed without merging**